third_party/pcsc-lite/naclport/js_client/src/reader-tracker-through-pcsc-api.js

```diff
@@ -0,0 +1,435 @@
+        // happened during previous call (see the SCardGetStatusChange docs for
+        // the details).
+        const currentState = readerStateOut['event_state'] & 0xFFFF;
+        return new API.SCARD_READERSTATE_IN(
+            readerStateOut['reader_name'], currentState);
+      });
+  // Add a magic entry that tells PC/SC to watch for the reader adding/removing
+  // events too (see the SCardGetStatusChange docs for the details).
+  readerStatesIn.push(new API.SCARD_READERSTATE_IN(
+      '\\\\?PnP?\\Notification', API.SCARD_STATE_UNAWARE));
+
+  this.logger_.fine(
+      'Waiting for the reader statuses change from PC/SC with the following ' +
+      'data: ' + GSC.DebugDump.dump(readerStatesIn) + '...');
+
+  api.SCardGetStatusChange(
+      sCardContext,
+      READER_STATUS_QUERY_TIMEOUT_MILLISECONDS,
+      readerStatesIn).then(function(result) {
+    result.get(function(readerStatesOut) {
+      this.logger_.fine('Received a reader statuses change event from PC/SC: ' +
+                        GSC.DebugDump.dump(readerStatesOut));
+      promiseResolver.resolve();
+    }, function(errorCode) {
+      if (errorCode == API.SCARD_E_TIMEOUT) {
+        this.logger_.fine('No reader statuses changes were reported by PC/SC ' +
+                          'within the timeout');
+        promiseResolver.resolve();
+      } else if (errorCode == API.SCARD_E_UNKNOWN_READER) {
+        this.logger_.warning(
+            'Waiting for the reader statuses changes from PC/SC finished ' +
+            'unsuccessfully due to removal of the tracked reader');
+        promiseResolver.resolve();
+      } else {
+        promiseResolver.reject(new Error(
+            'Failed to wait for the reader statuses change from PC/SC with ' +
+            'error code ' + GSC.DebugDump.dump(errorCode)));
+      }
+    }, this);
+  }, function(error) {
+    promiseResolver.reject(new Error(
+        'Failed to get the reader statuses from PC/SC: ' + error));
+  }, this);
+
+  return promiseResolver.promise;
+};
+
+/**
+ * Updates the result reported by this instance, given the reader states
+ * obtained from a call to the SCardGetStatusChange function.
+ * @param {!Array.<!API.SCARD_READERSTATE_OUT>} readerStates
+ * @private
+ */
+ReaderTrackerThroughPcscApi.prototype.updateResultFromReaderStates_ = function(
+    readerStates) {
+  this.updateResult_(goog.array.map(readerStates, function(readerState) {
+    return {
+      name: readerState['reader_name'],
+      isCardPresent: (readerState['event_state'] & API.SCARD_STATE_PRESENT) != 0
+    };
+  }));
+};
+
+/**
+ * Updates the result reported by this instance, and fires the update listener
+ * when necessary.
+ * @param {!Array.<{name: string, isCardPresent: boolean}>} result
+ * @private
+ */
+ReaderTrackerThroughPcscApi.prototype.updateResult_ = function(result) {
+  const isSame = goog.array.equals(
+      result, this.result_, goog.object.equals.bind(goog.object));
+  if (!isSame) {
```

Comment:
nit: This could just be if (isSame) return;

Reply:
Done